### PR TITLE
Added an article about deployments

### PR DESCRIPTION
**Pull request by zen-dog**

### CODE REVIEW

File: docs/articles/deployments.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+```
+{
+    "version": "2018-12-05T00:28:00.584Z",
+    "deploymentId": "342a6dc5-f871-4d10-a4ec-bbe21b431d58"
+}
+```
+We see that another `deploymentId` is returned. This is because internally canceling a deployment simply creats another deployment that reverts the steps of the original one.
+
+**Known caveats:** Marathon has a cap on the number of concurrently running deployments. This is controlled by `--max_running_deployments` [command line argument](command-line-arguments.md). In some rare occasions, one can have the maximum number of running deployments achieved so that canceling one is not possible (since a new deployment has to be created). In this case, canceling using `?force=true` parameter aka. [forced deployment](deployments.md#forced-deployments) should be used.
```
Comment (meln1k) [inline docs/articles/deployments.md]:
```suggestion
**Known caveats:** Marathon has a cap on the number of concurrently running deployments. This is controlled by `--max_running_deployments` [command line argument](command-line-arguments.md). In some rare cases it's possible to reach this limit so that canceling a deployments is not possible (since a new deployment has to be created). In this happens, canceling using `?force=true` parameter aka. [forced deployment](deployments.md#forced-deployments) should be used.
```

File: docs/articles/deployments.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+## Forced deployments
+Let's get back to our `product/service/rails-app` service deployment that is now "stuck" with slightly exaggerated resource requirements. We realize that 1k CPUs is too much and decide to update the `product/service/rails-app` service definition with reasonable CPU requirements. However, when submitting new service definition, Marathon will respond with `HTTP 209 (conflict)`:
```
Comment (meln1k) [inline docs/articles/deployments.md]:
```suggestion
Let's get back to our `product/service/rails-app` service deployment which is now "stuck" with slightly exaggerated resource requirements. We realize that 1k CPUs is too much and decide to update the `product/service/rails-app` service definition with reasonable CPU requirements. However, when submitting the new service definition, Marathon will respond with `HTTP 209 (conflict)`:
```

File: docs/articles/deployments.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+{
+    "message": "App is locked by one or more deployments. Override with the option '?force=true'",
+    "deployments": [
+        {
+            "id": "3aa7316a-7888-427b-9af3-4179a7791c79"
+        }
+    ]
+}
+```
+This is due to already existing (stuck) deployment that affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit our corrected definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). This time, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old ones if necessary and start new tasks.
```
Comment (meln1k) [inline docs/articles/deployments.md]:
```suggestion
This happened because already existing (stuck) deployment affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.
```

File: docs/articles/deployments.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+**Known caveats:** Due to its "destructive potential" usage of forced deployments is discouraged. A forced deployment will cancel any existing conflicting deployments (there might be more than one, see [group deployments](deployments.md#group-deployments)) potentially killing more tasks than the user intended.
+
+## Stopping a deployment
+In some rare cases a deployment might be stuck and canceling it is not an option (maybe because some tasks were already partially started and the user doesn't want to kill them). In this case, a deployment might be stopped. Stopping a deployment (as opposed to canceling a deployment) **will not return Marathon to the previous state**. Whichever tasks were already started will continue to run, the rest will not be launched. This operation will leave the service in an inconsistent state e.g. only 3/5 running tasks or 3 tasks running with the newest service definition version and 2 with the old one. **Stopping deployments should be a measure of last resort. Its usage is strongly discouraged**.
```
Comment (meln1k) [inline docs/articles/deployments.md]:
```suggestion
In some rare cases a deployment might be stuck and canceling it is not an option (for example because some tasks were already partially started and the user doesn't want to kill them). In this case, the deployment might be stopped. Stopping a deployment (as opposed to canceling a deployment) **will not return Marathon to the previous state**. Whichever tasks were already started will continue to run, the rest will not be launched. This operation will leave the service in an inconsistent state e.g. only 3/5 running tasks or 3 tasks running with the newest service definition version and 2 with the old one. **Stopping deployments should be a measure of last resort. Its usage is strongly discouraged**.
```

File: docs/articles/deployments.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+# Group deployments
+Let's assume that we have already deployed databases in our cluster and now it's time to deploy the services:
```
Comment (meln1k) [inline docs/articles/deployments.md]:
```suggestion
Let's assume that we already deployed the databases in our cluster and now it's time to deploy the services:
```

File: docs/articles/deployments.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+```
+/database
+    /mysql
+    /mongo
+```
+Note that we got rid of the top level `/product` group (the reason for this in the caveats section below).
```
Comment (meln1k) [inline docs/articles/deployments.md]: An anchor link would be really nice here.

File: docs/articles/deployments.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
+# Deployments
+
+# What is a deployment
+Every time a user creates a new [service](services.md) modifies or removes an existing one a deployment is created. Basically, deployment is a series of steps that Marathon has to execute in order to fulfill the user request. Let's consider an example where a few services are already deployed and running:
```
[the remark this answers is not in the document]
Comment (zen-dog) [reply docs/articles/deployments.md]: I submit that we should stick to [US spelling](https://www.grammarly.com/blog/fulfil-fulfill/) so it's _fullfill_ with double _ll_

File: docs/articles/deployments.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
+# Deployments
+
+# What is a deployment
```
[the remark this answers is not in the document]
Comment (zen-dog) [reply docs/articles/deployments.md]: that's taken from the template, so I guess we should keep it consistent?

File: docs/articles/deployments.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+- `StartApplication` - this step exists for backward compatibility reasons. Currently, it is doing nothing and is immediately successful
+- `ScaleApplication` - is responsible for starting the defined number of tasks and waiting for them to be up and running
+- `StopApplication` - stops given service by killing its tasks and waiting for the kill confirmation
+- `RestartApplication` - used when modifying an existing service definition and restarting its tasks (using the new definition)
+
+Steps are executed sequentially, however actions within one step (should there be more than one) are executed in parallel. Once all steps are executed successfully, the deployment is finished and removed from the state.
+
+## Canceling a deployment
+Sometimes a deployment is "stuck" for some reason (see [troubleshooting](troubleshooting.md) for more information). In this case, it might be helpful to cancel the deployment, returning to the previous state. Let's cancel the above stuck deployment by sending a `DELETE /v2/deployments/3bcb15fc-737f-4a36-b185-2ae3a703239e` request. The response is `HTTP 200(ok)` but the payload might be somewhat a surprise:
```
[the remark this answers is not in the document]
Comment (zen-dog) [reply docs/articles/deployments.md]: You can end up **without** the service after cancelation

File: docs/articles/deployments.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
+# Deployments
+
+# What is a deployment
+Every time a user creates a new [service](services.md), modifies or removes an existing one, new deployment is created. Basically, the deployment is a series of steps that Marathon has to execute in order to fulfil the user request. Let's consider an example where a few services are already deployed and running:
```
Comment (jeschkies) [inline docs/articles/deployments.md]:
```suggestion
Every time a user creates a new [service](services.md), modifies or removes an existing one, a new deployment is created. Basically, the deployment is a series of steps that Marathon has to execute in order to fulfill the user request. Let's consider an example where a few services are already deployed and running:
```

File: docs/articles/deployments.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+```
+{
+  "id": "/product/service/rails-app",
+  "cpus": 0.3,
+  "instances": 1,
+  "mem": 2048,
+  "cmd": ...
+}
+```
+If the provided definition is valid and there are no [deployment conflicts](deployments.md#forced-deployments) Marathon will return a `HTTP 201 (created)` response with a normalized service definition (including default fields that are allowed to be omitted) with an extra `deployments` field containing the deployment Id:
```
Comment (jeschkies) [inline docs/articles/deployments.md]: If I'm not mistaken we also set the deployment id in an extra header. Not sure if this is worth mentioning.

File: docs/articles/deployments.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+  "id": "/product/service/rails-app",
+  ...
+  "deployments":[{
+      "id": "3bcb15fc-737f-4a36-b185-2ae3a703239e"
+    }
+  ],
+```
+This deployment Id is a unique identifier of the process deploying the application. During the deployment Marathon will:
+
+- wait for Mesos to offer resources
```
Comment (jeschkies) [inline docs/articles/deployments.md]: Do we have articles for `offer` and `resource`?

File: docs/articles/deployments.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+In the above example `steps` field holds the deployment plan. Our deployment plan has two deployment steps: `StartApplication` and `ScaleApplication`. Currently, following deployment steps exist:
+
+- `StartApplication` - this step exists for backward compatibility reasons. Currently, it is doing nothing and is immediately successful
+- `ScaleApplication` - is responsible for starting the defined number of tasks and waiting for them to be up and running
+- `StopApplication` - stops given service by killing its tasks and waiting for the kill confirmation
+- `RestartApplication` - used when modifying an existing service definition and restarting its tasks (using the new definition)
+
+Steps are executed sequentially, however actions within one step (should there be more than one) are executed in parallel. Once all steps are executed successfully, the deployment is finished and removed from the state.
+
+## Canceling a deployment
```
Comment (jeschkies) [inline docs/articles/deployments.md]: Nit: Do we want to capitalize headers? So `Canceling a Deployment`?

File: docs/articles/deployments.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+{
+    "message": "App is locked by one or more deployments. Override with the option '?force=true'",
+    "deployments": [
+        {
+            "id": "3aa7316a-7888-427b-9af3-4179a7791c79"
+        }
+    ]
+}
+```
+This happened because already existing (stuck) deployment affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.
```
Comment (jeschkies) [inline docs/articles/deployments.md]: `existing, in our case the stuck deployment, affects our service definition`

File: docs/articles/deployments.md (hunk 11 above)
Comment (zen-dog) [reply docs/articles/deployments.md]: Good point! I'll add a sentence about it.

File: docs/articles/deployments.md (hunk 12 above)
Comment (zen-dog) [reply docs/articles/deployments.md]: Nope

File: docs/articles/deployments.md (hunk 13 above)
Comment (zen-dog) [reply docs/articles/deployments.md]: I thought about it, but as far as the template goes, we don't. Should we change it? Opinions?

File: docs/articles/deployments.md (hunk 14 above)
Comment (zen-dog) [reply docs/articles/deployments.md]: If you'd used the `Suggestion` feature, I could've accepted it in the browser ;)

File: docs/articles/README.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# This folder is work in progress
+
+In an [ongoing initiative](https://jira.mesosphere.com/browse/MARATHON-8518), we are working on a complete overhaul of the Marathon documentation. Articles in this folder will not be published until we have covered the majority of concepts.
```
Comment (alenkacz) [inline docs/articles/README.md]: 👍

File: docs/articles/deployments.md [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+{
+    "message": "App is locked by one or more deployments. Override with the option '?force=true'",
+    "deployments": [
+        {
+            "id": "3aa7316a-7888-427b-9af3-4179a7791c79"
+        }
+    ]
+}
+```
+This happened because existing, in our case the stuck deployment, affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.
```
Comment (alenkacz) [inline docs/articles/deployments.md]: link to group deployments from `service group definition`?

Also should we maybe link what `reconcile` means?

File: docs/articles/deployments.md [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+
+**Known caveats:** Due to its "destructive potential" usage of forced deployments is discouraged. A forced deployment will cancel any existing conflicting deployments (there might be more than one, see [group deployments](deployments.md#group-deployments)) potentially killing more tasks than the user intended.
```
Comment (alenkacz) [inline docs/articles/deployments.md]: if we recommend this , what is a recommended way to get out of the scenario described ^^^? If you have app that has e.g. resources so that it will never start? Should that just DELETE the deployment and not force override it?

File: docs/articles/deployments.md [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+## Stopping a Deployment
```
Comment (alenkacz) [inline docs/articles/deployments.md]: Maybe explanation of how does one stop a deployment?

File: docs/articles/instances.md [hunk 19]
```diff
@@ -0,0 +1,33 @@
+# Instances
+
+## What is an Instance?
+Instances are the Marathon representation of [⇒Tasks](tasks.md) in Mesos. An Instance can exist even if there is currently no Task associated. Depending on the goal of a given Instance, Marathon will take actions to reach that goal: it might match offers in order to launch a Task or [⇒TaskGroup](task-groups.md), or it might send kill requests to Mesos and eventually expunge the Instance from its state.
+
+Currently, instance goals are only used internally, but are reflected in the API. Supported goals are
```
Comment (alenkacz) [inline docs/articles/instances.md]: what does it mean that they are "reflected in the api"?

File: docs/articles/instances.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+## Which related concepts should I understand?
+When upgrading a Service during a [⇒Deployment](deployments.md) from one version to another, Marathon can create more instances than specified in either Service definition. It will do so if allowed via the [⇒upgradeStrategy](upgrade-strategy.md) configuration.
+When an instance is running on an agent disconnected from the cluster, it will eventually be reported as unreachable. According to the Service’s [⇒unreachableStrategy](unreachable-strategy.md), Marathon can schedule a replacement, even though the unreachable instance might still be running. When an unreachable instance becomes becomes reachable again, Marathon will scale down accordingly to the required amount of instances. It will decide over which instance shall be killed based on the [⇒killSelection](kill-selection.md).
+
+## How should I use Instances?
+_(This section could point out how to use Service definitions so that the user doesn’t need one app per instance, thus circumventing known limitations.)_
```
Comment (alenkacz) [inline docs/articles/instances.md]: I guess this will be done in a different PR?
